internal/expr/query.go

```diff
@@ -76,67 +75,95 @@ func (pe *PreparedExpr) Query(args ...any) (ce *QueryExpr, err error) {
 
 	// Query parameteres.
 	qargs := []any{}
-
-	for i, in := range pe.inputs {
-		v, ok := typeValue[in.structType]
+	for i, typeMember := range pe.inputs {
+		v, ok := typeValue[typeMember.outerType()]
 		if !ok {
 			if len(typeNames) == 0 {
-				return nil, fmt.Errorf(`type %q not passed as a parameter`, in.structType.Name())
+				return nil, fmt.Errorf(`type %q not passed as a parameter`, typeMember.outerType().Name())
 			} else {
-				return nil, fmt.Errorf(`type %q not passed as a parameter, have: %s`, in.structType.Name(), strings.Join(typeNames, ", "))
+				return nil, fmt.Errorf(`type %q not passed as a parameter, have: %s`, typeMember.outerType().Name(), strings.Join(typeNames, ", "))
+			}
+		}
+		switch tm := typeMember.(type) {
+		case structField:
+			qargs = append(qargs, sql.Named("sqlair_"+strconv.Itoa(i), v.FieldByIndex(tm.index).Interface()))
+		case mapKey:
+			// MapIndex returns a zero value if the key is not in the map so we
+			// need to check the MapKeys().
+			var val reflect.Value
+			for _, key := range v.MapKeys() {
+				if key.String() == tm.name {
```

Comment:
It does, but as far as I've been able to work out this is the only way to ensure that a key is in the map.

As explained in the comment above this loop, if we use `MapIndex` on the reflected value a zero value is returned when the key is not in the map which is indistinguishable from a key that is in the map and is set to zero.

It would be good if we could simply use it as a map, rather than a reflected value but then we run into the problem of not knowing the maps value type. We allow people to pass `map[string]any` or `map[string]int` ect.. We cant index the map without knowing its type and as far as I know there is no way to cast to a specific concrete type at runtime with reflect (without enumerating all possibilities).

We could of course simple set the argument to the zero value, but to me that would feel like creating an easy way to introduce errors (for example a spelling mistake in the key name would go undetected).

Reply:
Changed to use `MapIndex`. It turned out it returns a `reflect.Invalid` if it cannot find the map key instead of a zero value the map value type.